### ISSUE TITLE
Errors while trying to run this using .NET 6 arm64 SDK with M1 Macbook Pro

### ISSUE THREAD

**Issue by Minorsonek** [confluentinc/confluent-kafka-dotnet#1755]
Description
===========

Im getting an error when trying to use this inside .net 6 web project on M1 macbook pro (arm64 architecture). The error is:
 System.DllNotFoundException: Failed to load the librdkafka native library.
         at Confluent.Kafka.Impl.Librdkafka.TrySetDelegates(List`1 nativeMethodCandidateTypes)
         at Confluent.Kafka.Impl.Librdkafka.LoadOSXDelegates(String userSpecifiedPath)
         at Confluent.Kafka.Impl.Librdkafka.Initialize(String userSpecifiedPath)
         at Confluent.Kafka.Consumer`2..ctor(ConsumerBuilder`2 builder)
         at Confluent.Kafka.ConsumerBuilder`2.Build()

It works if i run the project on older .net5 using rosetta, without any code change. If i update to .net6 and use arm, it gives this error on first consumer/producer usage.
I have tried side loading the library from "brew install librdkafka" using Library.Load, but no success there. Any chance to get it to work on .net 6, or is it not supported yet?

How to reproduce
================
I think description says it all, just try to install only .net6 sdk on m1 macbook and try to run this nuget

Checklist
=========

Please provide the following information:
 - [x ] A complete (i.e. we can run it), minimal program demonstrating the problem. No need to supply a project file.
 - [ 1.8.2] Confluent.Kafka nuget version.
 - [1.8.2] Apache Kafka version.
 - [ x] Client configuration.
 - [ MacOs Monterey] Operating system.
 - [description ] Provide logs (with "debug" : "..." as necessary in configuration).
 - [ x] Provide broker log excerpts.
 - [ for me, yes, for others, probably not so critical] Critical issue.

**Comment by k-kapsazov** [comment]
Hi,
The other day my team hit the same problem. After some digging in the code we found the actual error is:

```
Unable to load shared library 'librdkafka' or one of its dependencies. In order to help diagnose loading problems, 
consider setting the DYLD_PRINT_LIBRARIES environment variable:
 dlopen(liblibrdkafka, 0x0001): tried: 'liblibrdkafka' (no such file), 
 '/usr/local/lib/liblibrdkafka' (no such file), 
 '/usr/lib/liblibrdkafka' (no such file), 
 '/Users/<your user>/<your project>/bin/Debug/net6.0/liblibrdkafka' (no such file), 
 '/usr/local/lib/liblibrdkafka' (no such file), 
 '/usr/lib/liblibrdkafka' (no such file)
```

So we concluded that the actual package is missing from the right place.

So as a workaround you can do the following:

1. run command `brew install librdkafka`

2. copy all files from `/opt/homebrew/Cellar/librdkafka/1.8.2/lib` to one of the suggested location in the error (We choose the directory of the project '/Users/<your user>/<your project>')

3. mark all added files as "Content" and "Copy always" in your IDE so they get transferred to the `/bin/Debug` folder on build

4. after build everything should work without an error

PS: We tried setting the working directory of librdkafka with the line `Library.Load("/opt/homebrew/Cellar/librdkafka/1.8.2/lib")` but it didn't work.

**Comment by vijaybw** [comment]
@k-kapsazov wow, I am glad the steps worked for you. I am facing the same issue but the steps did not work for me.

Do you mind sharing the complete path of the location where you kept files inside lib folder? Also, do you maintain the hierarchy of the files insider 1.8.2/lib inside /bin/Debug folder?

Also, any other tips will also help. I have Mac pro with M1.

/usr/local/share/dotnet/dotnet /Users/myprj/bin/Debug/net6.0/myprj.Web.dll
Unhandled exception. System.DllNotFoundException: Failed to load the librdkafka native library.
   at Confluent.Kafka.Impl.Librdkafka.TrySetDelegates(List`1 nativeMethodCandidateTypes)
   at Confluent.Kafka.Impl.Librdkafka.LoadOSXDelegates(String userSpecifiedPath)
   at Confluent.Kafka.Impl.Librdkafka.Initialize(String userSpecifiedPath)
   at Confluent.Kafka.Consumer`2..ctor(ConsumerBuilder`2 builder)
   at Confluent.Kafka.ConsumerBuilder`2.Build()

**Comment by vijaybw** [comment]
Thank you for the response. The above fix help me run the application without error.
But, I get error when my producers starts sending a message.

Unhandled exception. Confluent.Kafka.ProduceException`2[System.String,System.String]: Local: Invalid argument or configuration
   at Confluent.Kafka.Producer`2.ProduceAsync(TopicPartition topicPartition, Message`2 message, CancellationToken cancellationToken)
   at Producer.Implementations.SampleProducer.PublishAsync(String topic, String message) in /Users/cuser/Downloads/Start-with-Kafka-DotNetCore-master/Producer/Implementations/SampleProducer.cs:line 21
   at Producer.Program.Main(String[] args) in /Users/cuser/Start-with-Kafka-DotNetCore-master/Producer/Program.cs:line 19
   at Producer.Program.<Main>(String[] args)

I followed couple of different tutorials but I get the same message. https://medium.com/@shesh.soft/confluent-kafka-integration-with-net-core-2a489fa2512

Any clue?

**Comment by artem-fedorov** [comment]
I got the same issue with M1, confluent Kafka, and .Net 6.  Do you have any ETA for resolving this issue?

**Comment by MaorDavidzon** [comment]
+1

**Comment by vijaybw** [comment]
https://github.com/confluentinc/confluent-kafka-dotnet/issues/1769 could be the fix for this one.

**Comment by mhowlett** [comment]
closing in favor of #1769 - thanks for reporting.

**Comment by rrehman-hbk** [comment]
> "Local: Invalid argument or configuration" when building as 64 bit #949

I had the same issue. Then I had to uninstall .NET Core sdks and runtimes and installed Visual Studio (so using .NET Core sdks from Visual Studio). This solved my issue. Not sure if this helps you.

**Comment by alexknips** [comment]
> > "Local: Invalid argument or configuration" when building as 64 bit #949
> 
> I had the same issue. Then I had to uninstall .NET Core sdks and runtimes and installed Visual Studio (so using .NET Core sdks from Visual Studio). This solved my issue. Not sure if this helps you.

Not really. I am also using Rider on Mac and this is not possible.

**Comment by rrehman-hbk** [comment]
> installed Visual Studio

https://visualstudio.microsoft.com/vs/mac/
I used this vscode, before installing removed all .NET code sdks and runtimes. Then installed Rider. So Rider was using the .NET sdks provided by vscode (which choses the SDKs from intel and apple as per the requirement as per my understanding)

**Comment by jameswmoar** [comment]
For the benefit of anyone coming across this thread now, this issue was apparently resolved in v1.9.2 of Confluent.Kafka. I've confirmed that it is definitely resolved in v2.1.1.
